scan_packages: made adding package managers easier

bcoca opened the pull request:
now it can scan for multiple package managers per system
  also added portage support

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
scan_packages

bcoca committed:
made adding package managers easier

  added portage support

webknjaz commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -141,93 +146,174 @@
+
+    # get supported pkg managers
+    #  TODO: make this dynamic PKG_MANAGERS = [getattr(x, 'name') for x in inspect.getmembers(me) if inspect.isclass(x) and hasattr(x, 'name')]
+    PKG_MANAGERS = ['apt', 'rpm', 'portage']
```
OS-level only? What about language-based?

webknjaz commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -141,93 +146,174 @@
+    def list_installed(self):
+        rc, out, err = self.m.run_command(' '.join([self.cli, '-Iv', '|', 'xargs', '-n', '1024', 'qatom']), use_unsafe_shell=True)
+        if rc != 0 or err:
+            raise Exception("Unable to list packages rc=%s : %s" % (rc, err))
```
```suggestion
            raise RuntimeError("Unable to list packages rc=%s : %s" % (rc, to_native(err)))
```

webknjaz commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -141,93 +146,174 @@
+from abc import abstractmethod
 
 from ansible.module_utils.basic import AnsibleModule
 from ansible.module_utils._text import to_text
```
```suggestion
 from ansible.module_utils._text import to_native, to_text
```

webknjaz commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -19,11 +19,16 @@
+      - The package manager used by the system so we can query the package information.
+      - Since 2.8 this is a list and can support multiple package managers per system.
+      - The 'portage' option was added in version 2.8.
+    default: ['auto']
```
Same style please:
```suggestion
    default: ["auto"]
```

webknjaz replied:
@bcoca ^

webknjaz commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -141,93 +146,174 @@
+class PORTAGE(CLIMgr):
+
+    name = 'portage'
+    CLI = 'qlist'
```
Can we use `eix` when it's present?

bcoca replied:
yes and no, eix-installed gives similar info but not the same, i can make a conditional branch but the results will differ.

webknjaz commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -141,93 +146,174 @@
-
-
-# FIXME: add more listing methods
+class PkgMgr(object):
```
```suggestion
class PkgMgr:
```

webknjaz commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -141,93 +146,174 @@
+        pass
+
+    @abstractmethod
+    def get_package_details(self):
```
Signature of the abstract method doesn't match the implementation and calls
```suggestion
    def get_package_details(self, package):
```

webknjaz commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -141,93 +146,174 @@
+
+    @abstractmethod
+    def get_package_details(self):
+        pass
```
Maybe?
```suggestion
        raise NotImplementedError
```

abadger replied:
This is an @abstractmethod so we do not want to raise NotImplementedError.

* abstractmethods will fail when a class is instantiated whereas NotImplementedError is triggered when the method is called
* abstractmethod can be used in the usual way: ```super(MyClass, self).method()``` which will run any code that is present in the abstractmethod.  If it has NotImplementedError, then that will error.

webknjaz replied:
Right, makes sense.

webknjaz commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -141,93 +146,174 @@
+        fields = package.split()
+        for i, value in enumerate(fields):
+            pkg[self.atoms[i]] = value
+        return pkg
```
```suggestion
        return dict(zip(self.atoms, package.split()))
```

webknjaz commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -141,93 +146,174 @@
+        pkg = {}
+        fields = package.split()
+        for i, value in enumerate(fields):
+            pkg[self.atoms[i]] = value
```
```suggestion
```

webknjaz commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -141,93 +146,174 @@
+
+        pkg = {}
+        fields = package.split()
+        for i, value in enumerate(fields):
```
```suggestion
```

webknjaz replied:
If you still want explicit loop it'd be:
```suggestion
        for key, value in zip(self.atoms, fields):
                pkg[key] = value
```

webknjaz replied:
Or

```suggestion
        return dict((key, value) for key, value in zip(self.atoms, fields))
```

webknjaz commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -141,93 +146,174 @@
+    def get_package_details(self, package):
+
+        pkg = {}
+        fields = package.split()
```
```suggestion
```

webknjaz commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -141,93 +146,174 @@
+
+    def get_package_details(self, package):
+
+        pkg = {}
```
```suggestion
```

bcoca committed:
moar pkg mgrs and moar info

 - added 'pkg' pkg manager (freebsd)
 - added pip
 - more apt info

bcoca committed:
updated clgo

webknjaz committed:
Updates from feedback

Co-Authored-By: bcoca <bcoca@users.noreply.github.com>

bcoca committed:
incorporated more feedback and added docstrings

abadger commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -140,94 +145,236 @@
+
+        self.m = module
+
+    @abstractmethod
```
This decorator has no effect without also using the proper metaclass.

abadger commented:
This makes it hard to program against the return values as the fields as heterogeneous using this.  Need to think of a better structure for the return values so that we only have packages with the same fields together.

abadger commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -140,94 +145,236 @@
+
+def main():
+
+    me = sys.modules[__name__]
```
You don't need this.  Use ```globals().get([..])``` instead.

webknjaz replied:
@abadger you can suggest changes by hitting Ctrl+G ;)
```suggestion
    me = globals().get(__name__)
```

bcoca replied:
I'm unsure what you mean here, can you expand?

abadger replied:
here I meant:
* remove this line ```me = sys.modules[__name__]``` 
* When you retrieve the manager do it like this:             ```manager = globals().get(pkgmgr.upper())```

However, we shouldn't have to do either of these if you use ```get_all_subclasses()``` below.  We can do it in a better way then as PKG_MANAGERS will be a dict mapping directly to the class we need.

bcoca replied:
dont need this anymore due to pkg_mangers updates below

bcoca commented:
I was thinking of enforcing a 'minimal' structure (mandatory: 'name, version, source') and leaving the rest variable as the different systems provide data that might interest some and not others, but even then try to normalize as much as possible (optional: category, origin, auto, etc).

bcoca committed:
moar from feedback

  - made manager list dynamic and names based on class
  - better not found msg
  - made abstract metaclass again
  - test is now init exception
  - module to global
  - better dedupe comments

abadger commented:
I think you were right to  use @abstractmethod.  You just have to use it correctly.  (The toplevel class in the hierarchy should be using six.with_metaclass=ABCMeta)

abadger commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -140,94 +145,231 @@
+
+    # get supported pkg managers
+    #  TODO: make this dynamic PKG_MANAGERS = [getattr(x, 'name') for x in inspect.getmembers(me) if inspect.isclass(x) and hasattr(x, 'name')]
+    PKG_MANAGERS = ['apt', 'rpm', 'portage', 'pkg', 'pip']
```
I think this will work:

``` python
PKG_MANAGERS = dict((mgr.__name__.lower(), mgr) for mgr for mgr in basic.get_all_subclasses(PKGMGR) if mgr not in (CLIMgr, LibMgr))
```

EDIT: Used the objects instead of strings

bcoca replied:
done

abadger commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -140,94 +145,231 @@
+
+        self.m = module
+
+    def test(self):
```
Also mark the test methods as @staticmethod.

That way the code won't have to instantiate a class until after it's been shown that the object is needed.

It's probably even better to have ```__init__()``` run test(), though, throw an exception when the necessary pieces to use that class are not found, and then have the caller catch those exceptions.

bcoca replied:
done

abadger commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -140,94 +145,231 @@
+
+    def test(self):
+        ''' This method is supposed to return True/False if the package manager is installed/usable by this module '''
+        raise NotImplementedError
```
When you use ABCMeta and abstractmethod, you should just have these ```pass```.  That way an implementor can call the base class's method if they want and it won't raise an exception.

bcoca replied:
i had reverted from that as per previous suggestions, any drawbacks of doing it with ABCMeta?

abadger commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -140,94 +145,231 @@
+            package_details = self.get_package_details(package)
+            package_details['source'] = self.name
+            name = package_details['name']
+            if name not in installed_packages:
```
Ideal place to use ```collections.defaultdict``` ;-)

bcoca replied:
for some reason i keep avoiding it .. yet it was added in 2.5 ...

abadger commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -140,94 +145,229 @@
+    def test(self):
+        works = super(APT, self).test()
+        if works:
+            self._cache = self.lib.Cache()
```
This isn't part of the test so it shouldn't be done here.  For the way you want to use it, it's probably best done as an @property.
``` python
@property
def pkg_cache(self):
    if self._cache:
        return self._cache

    self._cache = self.lib.Cache()
    return self._cache
```

abadger replied:
Then, in the rest of your code, change ```self._cache``` into ```self.pkg_cache```

bcoca replied:
done

abadger commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -140,94 +145,229 @@
+
+    def get_package_details(self, package):
+        ac_pkg = self._cache[package].installed
+        return dict(name=package, version=ac_pkg.version, arch=ac_pkg.architecture, category=ac_pkg.section, origin=ac_pkg.origins[0].origin)
```
MINOR: i think we decided to prefer literals over calling dict().

bcoca replied:
mostly copy/paste from original, i'm fine either way

webknjaz replied:
let's stick to literals, please

abadger commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -140,94 +145,229 @@
 
-    if manager is None or manager == 'auto':
+    def get_package_details(self, package):
+        return package
```
This needs to return a dict and needs to return version as well as package name.

bcoca replied:
which it should be for this plugin, which does a json.loads, we can add verification, but i would do that in parent then and use super on all children?

abadger replied:
That would be a bug in list_installed(), then.  We should be returning the same data types across all the plugins. So list_installed() should be returning a list of package names.

(Probably want to use a @property here as well to cache the json.loads() output)

bcoca replied:
no i mostly designed as 2 functions for one to query the list of packages and the other to properly format, in the case of pip with `--format=json` it all already happens in the first function

abadger commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -140,94 +145,229 @@
+    PKG_MANAGERS = dict([(obj.__name__.lower(), obj) for obj in get_all_subclasses(PkgMgr) if obj not in(CLIMgr, LibMgr)])
+
+    # start work
+    module = AnsibleModule(argument_spec=dict(manager={'type': 'list', 'default': ['auto']}), supports_check_mode=True)
```
Rather than passing the module into each PkgMgr class, it makes more sense to leave this as a global.  The classes only need AnsibleModule for two utility functions which have nothing to do with the PkgMgr classes themselves.

abadger commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -140,94 +145,229 @@
+
+    if 'auto' in managers:
+        managers.extend(PKG_MANAGERS.keys())
+        managers.remove('auto')
```
This will lead to duplicates.  It will be better to do:
``` python
if 'auto' in managers:
    managers = list(PKG_MANAGERS.keys())
```

abadger replied:
Ah... I see that you want to preserve order.  You should take the comment from below and add it here.

bcoca replied:
i want to let user set order, i do dedupe later on

abadger commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -140,94 +145,229 @@
+        managers.extend(PKG_MANAGERS.keys())
+        managers.remove('auto')
+
+    unsupported = set(managers).difference(set(PKG_MANAGERS.keys()))
```
Note: you don't have to call set() explicitly the second time, nor do you have to call keys() explicitly:

``` python
unsupported = frozenset(managers).difference(PKG_MANAGERS)
```

bcoca replied:
understood, but being explicit i don't need to think about what happens under the covers implicitly

abadger commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -140,94 +145,229 @@
+                packages.update(manager.get_packages())
+        except Exception as e:
+            from traceback import format_tb
+            results['errors'].append({'msg': 'Failed to retrieve packages: %s' % to_text(e), 'exception': format_tb(sys.exc_info()[2])})
```
There's some cornercases here which make me think we want to determine whether this is a warning or an error later.  A user can install package management software on any box but it might not work on that box.  That might not make a difference to the user at all.

(because it's not the primary package manager on the system.  For instance, they just needed it to locally examine the API of a package manager on a foreign system.)

bcoca replied:
here i expect to find all kinds of corner cases, which is why i ignore this mostly by default, using warnings in stead might be fine

abadger commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -140,94 +145,229 @@
+            from traceback import format_tb
+            results['errors'].append({'msg': 'Failed to retrieve packages: %s' % to_text(e), 'exception': format_tb(sys.exc_info()[2])})
+
+    if found == 0:
```
If we just use found for the purpose of this error, then it should be a boolean.  We might want to do something more complex with errors, though, in which case we'd want it to be something more than either a boolean or a simple count.

bcoca replied:
i now distinguish between test failure and other errors, so this should show as warnings or errors depending on which path they hit

abadger commented:
Three global thoughts:

* For the return values, I think that a minimal structure has benefits and would be sufficient if the module wasn't returning values from multiple package managers in the same invocation.  However, since you're now having the module return packages from multiple package managers, it no longer is sufficient.  A person can no longer iterate the list of packages and know what fields they are going to encounter, even in an environment which has heterogeneous platforms.  So if we're going to allow this we need to separate the values that differ per package manager.  Maybe pip_package[], rpm_package[], etc (possibly with a return value that combines all of them in addition?)

* Errors, there's a few cornercases around errors.  Trying to step back far enough to find a general rul... perhaps we want to return errors if anything that the user specified as an explicit package manager type failed and return warnings for anything that was pulled in via auto?

* non-system package managers: If we expand this to non-system package managers like pip, then we have multiple package managers of *a single type*.  For instance, we'd need to account for /usr/bin/pip2.7 /usr/bin/pip3.6 and /usr/bin/pip3.7 all being installed on the same system.  Should we avoid non-system package managers in this module for the moment?

bcoca commented:
- For the return values, you always have 'souce' which will indicate the fields provided, very similar as what you propose. I left it as is to avoid changing the general format and those already using the module won't have to change anything

- The errors i'm still 'evolving' on how to handle, displaying only for explicit sounds good, we'll probably iterate more over this as we get user feedback

- I added pip mostly as example .. but it does have issues with different python interpreters and `--user flag`, also this would change the classification of the plugin itself `packaging/os` ... so i'm fine with dropping it for now or revisiting the interface (allow overriding of the cli searched?).

bcoca committed:
more targetted errors/warnings

bcoca committed:
added strategy, reordered to conserve priority

bcoca committed:
rpm > apt

bcoca committed:
move break to top

bcoca committed:
fix trate

abadger commented:
So there seems to be several changes in this single PR... here's what I think about each:
* Adding portage +1
* refactoring the code into modular pieces based on classes +1
* Adding the ability to read from multiple package managers -1
  * This feature would be okay if the return values were changed to separate the packages from different package managers into separate output fields for each manager.  But without that we'd be shoving nonheterogeneous output into a single return field which isn't the right way to do things.
  * sivel also raises the problem that strategy will cause us to need to namespace the results further.  I agree with that assessment for strategy.  Changing the return value to have separate fields for each manager would solve this problem as well as we could get rid of strategy altogether and return output from all managers everytime.

bcoca commented:
except that breaks backwards compatibility with existing playbooks, as is the defaults will still work as they did before. You only have collision issues if you keep mismatching lists, but that is up to the user and we can document it as a consideration, which already exists for most facts modules, including setup.

bcoca committed:
piping it

abadger commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -139,95 +150,256 @@
+
+class PkgMgr(object):
+
+    __metaclass__ = ABCMeta
```
The correct way to use metaclasses in code that needs to run on both python2 and python3 is:
``` python
from abc import ABCMeta
from six import with_metaclass


class PkgMgr(with_metaclass(ABCMeta, object)):
```

bcoca committed:
lines and meta

webknjaz commented on changelogs/fragments/scan_packages.yml:
```diff
@@ -0,0 +1,5 @@
+minor_changes:
+    - rewrote scan_packages for easier support of  multiple package managers per system
```
This is rendered in changelog, so should probably be "rewritten"

webknjaz commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -19,11 +19,22 @@
+    type: list
+  strategy:
+    description:
+      - How to query the package managers on the system
```
Please add the info on what those options actually mean.

webknjaz commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -19,11 +19,22 @@
+    version_added: "2.8"
 version_added: "2.5"
+requirements:
+    - For 'portage' support it requires the `qlist` utility, which is part of the 'portage-utils' package.
```
```suggestion
    - For 'portage' support it requires the `qlist` utility, which is a part of the 'app-portage/portage-utils' atom.
```

webknjaz commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -139,95 +150,275 @@
           }
         }
 '''
+import json
```
Let's keep an empty line separator:
```suggestion

import json
```

webknjaz commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -139,95 +150,275 @@
+from ansible.module_utils.six import with_metaclass
+
+
+class TestFailed(Exception):
```
AFAICS this exception is a `RuntimeError` semantically:
```suggestion
class TestFailed(RuntimeError):
```

webknjaz replied:
Also, the naming is odd. It's too generic and refers to software testing at the first look, misleading for the reader.
I suggest something more connected to its purpose:
```suggestion
class PackageManagerMissing(RuntimeError):
```
(or: s/Missing/Unavailable/)

webknjaz commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -139,95 +150,275 @@
+    pass
+
+
+class PkgMgr(with_metaclass(ABCMeta, object)):
```
`six.add_metaclass` is better: https://stackoverflow.com/a/37450110/595220.
Also, `__metaclass__ = type` as the top allows us to omit inheriting from the `object` explicitly:
```suggestion
@add_metaclass(ABCMeta)
class PkgMgr:
```

webknjaz replied:
Oh, and I'd like you see readable names. both pkg and mgr shortcuts have a lot of different meanings.
`PackageManager` would be more than enough:
```suggestion
@add_metaclass(ABCMeta)
class PackageManager:
```

abadger replied:
Ah, add_metaclass  became available when we dropped python-2.4 support (which enabled us to update to a version of six which contained it).  I believe object is needed in the with_metaclass form but not in the add_metaclass form.

webknjaz replied:
Yep. And according to stackoverflow it fixes some inheritance quirks.

webknjaz commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -139,95 +150,275 @@
+            raise TestFailed("Test failed to pass")
+
+    @abstractmethod
+    def test(self):
```
Let's use self-explanatory name for this method:
```suggestion
    def verify_cli_backend_available(self):
```

webknjaz commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -139,95 +150,275 @@
+
+
+class PkgMgr(with_metaclass(ABCMeta, object)):
+
```
Base should probably have a standardized attribute used in child classes:
```suggestion

    CLI = None

```

abadger replied:
I'd change them to abstractproperty:
``` python
@abstractattribute
def CLI(self):
    # This should simply be an attribute containing the UNIX command for this package manager.
    pass
```

This way any class which implements the interface will need to override it with a CLI attribute or property.

webknjaz replied:
Agreed. That's even better!

webknjaz commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -139,95 +150,275 @@
+    def __init__(self):
+
+        if not self.test():
+            raise TestFailed("Test failed to pass")
```
This looks like there's some testing in place, while it's just a conditional prerequisites check. It's much more friendly if you put some clue about what's happening there:
```suggestion
            raise PackageManagerMissing("%s is not installed or unavailable" % (self.CLI, ))
```

webknjaz commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -139,95 +150,275 @@
+
+    @abstractmethod
+    def test(self):
+        ''' This method is supposed to return True/False if the package manager is installed/usable by this module '''
```
It's nicer to read PEP257+PEP8 styled docstrings:
```suggestion
        """Identify whether the package manager is available to use.

        This method returns True if the package manager is installed and
        usable by this module. It contains a package-specific
        implementation of such check.
        """
```

webknjaz replied:
@bcoca?

webknjaz commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -139,95 +150,275 @@
+        pass
+
+    @abstractmethod
+    def get_package_details(self, package):
```
It is better to have consistent method naming: sometimes you use "package" in names, sometimes you don't.
I suggest reconsidering it and maybe omit word "package" at all. In this case, for example, you use it as an argument, which makes it obvious even when you change it to `get_details_of(self, package)` what it's about. Otherwise you can choose to use "package" in all related methods, consistency matters.

webknjaz commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -139,95 +150,275 @@
+    def get_packages(self):
+        ''' Take all of the above and return a dictionary of lists of dictionaries (package = list of installed versions) '''
+
+        installed_packages = {}
```
Please use:
```suggestion
        installed_packages = defaultdict(list)
```

This way you'll simplify logic below.

webknjaz commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -139,95 +150,275 @@
+        installed_packages = {}
+        for package in self.list_installed():
+            package_details = self.get_package_details(package)
+            package_details['source'] = self.__class__.__name__.lower()
```
You keep doing this on every loop iteration, which is inefficient. Please move it outside of the loop.

webknjaz commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -139,95 +150,275 @@
+        pass
+
+    @abstractmethod
+    def list_installed(self):
```
I think you don't really need a list but any iterable. And I'd prefer it to be a generator since you always iterate through this loop one time and then throw it away. It's more efficient to `yield` things or return a generator (expression).

webknjaz commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -139,95 +150,275 @@
-                break
-            except ImportError:
-                pass
+                # manager throws TestFailed exception on init (calls self.test) if not usable.
```
If you rebrand the exception this comment will be no longer needed.

webknjaz commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -139,95 +150,275 @@
+
+class LibMgr(PkgMgr):
+
+    LIB = None
```
Can we use a variable style for this rather then a constant? it looks off.
I'd also say `lib_name` to differentiate from `lib` (which might be `lib_module`).
When I read LIb LIB lib lIB I cannot really distinguish between those immediately.

webknjaz commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -139,95 +150,275 @@
+
+class LibMgr(PkgMgr):
+
+    LIB = None
```
Also, attributes deserve to have their own docstrings:
```suggestion
    lib_name = None
    """Name of importable used to access the current package manager."""
```

abadger replied:
Docstrings for attributes should be formatted as:
``` python
#: Import name of library for interacting with the package manager
@abstractproperty
def lib_name(self):
    pass
```

http://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#directive-autoattribute

webknjaz replied:
No, that's not attribute docstrings.

PEP 257 says:
> String literals occurring immediately after a simple assignment at the top level of a module, class, or __init__ method are called "attribute docstrings".

That's my intention. They look like in code and they shouldn't be comments.


Your own link to sphinx docs says that it supports this style as well:
> For module data members and class attributes, documentation can [...], **or in a docstring** _after_ the definition.

And then it gives example:
```python
class Foo:
    """Docstring for class Foo."""

    [...]

    baz = 2
    """Docstring for class attribute Foo.baz."""

    def __init__(self):
        [...]

        self.spam = 4
        """Docstring for instance attribute spam."""
```

abadger replied:
Ah, I remember discussing htis with someone else many years ago....

Currently there are no attribute docstrings in python.  So my preference is that we use the comment form rather than the string literal form.  Comments pose a lower overhead to the interpreter (they are ignored while string literals are parsed, processed, and objects created) and since the string literal form is ultimately discarded because they aren't assigned to any variable names it makes sense to force python to do as little work as possible for these.

webknjaz commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -139,95 +150,275 @@
+            found = True
+        except ImportError:
+            pass
+        return found
```
It probably even makes more sense to put exception raising into this function.
Otherwise it's a three-step process on the high level: first, call the method, then, check the result, finally raise an exception.

It would be much cleaner to have smth like `self.verify_cli_backend_available()` in the initializer and that's it.

abadger replied:
+1

webknjaz commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -139,95 +150,275 @@
+
+class CLIMgr(PkgMgr):
+
+    CLI = None
```
Same comments from LibMgr apply here as well.

webknjaz commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -139,95 +150,275 @@
+    PKG_MANAGERS = dict([(obj.__name__.lower(), obj) for obj in get_all_subclasses(PkgMgr) if obj not in(CLIMgr, LibMgr)])
+
+    # start work
     global module
```
Why not initialize the module inside of PkgMgr? You wouldn't have to deal with globals then.

abadger replied:
PkgMgr doesn't have anything to do with module.  A global is the best fit for an AnsibleModule's scope.

webknjaz replied:
Well, you it has methods referring to it. How about passing an instance via initializer?

abadger replied:
As I said, PkgMgr has nothing to do with AnsibleModule.  They're two separate entities.  If you think of it conceptually, it also makes sense.  This whole file is "an ansible module".  So any object within the file really should be able to access the AnsibleModule object.  Using a global is a better model of the real world here.

webknjaz commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -139,95 +150,275 @@
+        return pkg
+
+
+class PORTAGE(CLIMgr):
```
Let's stick to CamelCase naming style, it's a class after all:
```suggestion
class Portage(CLIMgr):
```

webknjaz commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -139,95 +150,275 @@
+                pkg['port_epoch'] = 0
+
+            if '_' in pkg['version']:
+                pkg['version'], pkg['revision'] = pkg['version'].split('_', 1)
```
How about:
```suggestion
            pkg['version'], _, pkg['revision'] = pkg['version'].partition('_')
            if not pkg['revision']:
                pkg['revision'] = "0"
```

abadger replied:
Note, use dummy instead of _ for throwaway variables.

webknjaz replied:
Why?

abadger replied:
_ is the accepted alias for gettext functions.  So if we ever decide to i18n and l10n the code, we'll want to use _ for that meaning.  Local variables that override that will then cause unexpected bugs.

webknjaz commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -139,95 +150,275 @@
+            if '_' in pkg['version']:
+                pkg['version'], pkg['revision'] = pkg['version'].split('_', 1)
+            else:
+                pkg['revision'] = 0
```
You have int value here while the other branch assigns a string.

webknjaz commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -139,95 +150,275 @@
+
+    def list_installed(self):
+        rc, out, err = module.run_command(' '.join([self.cli, '-Iv', '|', 'xargs', '-n', '1024', 'qatom']), use_unsafe_shell=True)
+        if rc != 0 or err:
```
I'm not sure why stderr should indicate a error? It's meant for warnings and other non-parsable output.

bcoca replied:
was just inertia, soo many tools improperly use rc, but Gentoo devs have always been good about it.

webknjaz commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -139,95 +150,275 @@
 def main():
+
+    # get supported pkg managers
+    PKG_MANAGERS = dict([(obj.__name__.lower(), obj) for obj in get_all_subclasses(PkgMgr) if obj not in(CLIMgr, LibMgr)])
```
Could you please multiline this? Also an intermediate list is not needed:
```suggestion
    PKG_MANAGERS = dict(
        (obj.__name__.lower(), obj)
        for obj in get_all_subclasses(PkgMgr)
        if obj not in (CLIMgr, LibMgr)
    )
```

webknjaz commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -139,95 +150,275 @@
     global module
-    module = AnsibleModule(argument_spec=dict(manager=dict()), supports_check_mode=True)
-    manager = module.params['manager']
+    module = AnsibleModule(argument_spec=dict(manager={'type': 'list', 'default': ['auto']},
```
Can we have a dict literal, please?

webknjaz commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -139,95 +150,275 @@
-        from traceback import format_tb
-        module.fail_json(msg='Failed to retrieve packages: %s' % to_text(e), exception=format_tb(sys.exc_info()[2]))
+        except Exception as e:
+            if pkgmgr in module.params['manager']:
```
snake_case plz
```suggestion
            if pkg_mgr in module.params['manager']:
```

abadger commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -139,95 +150,275 @@
+
+
+class PIP37(PIP):
+    CLI = 'pip3.7'
```
I don't think this is maintainable.  We'll have to think of a different structure for handling things like pip.  Probably, as you suggested on slack, something revolving around passing in command line binaries explicitly.

mattclay replied:
Instead of invoking `pip` directly it should be invoked through the python interpreter, like:

```python
[sys.executable, '-m', 'pip', ...]
```

bcoca replied:
that would limit it to the 'current python used to execute module' while executing the pip version is independent of that.

webknjaz replied:
I'd be more precise.

bcoca replied:
in any case, i don't think it will end up living here but in a 'language' or 'pip_package's module, just works nicely for my systems right now

bcoca committed:
refactored common functions

 - moved pip into it's own module
 - cleaned up base clases
 - ensure 'lower' match in package_facts

bcoca committed:
missing license

bcoca committed:
avoid facts

bcoca committed:
update clog

bcoca committed:
addressed feedback

bcoca committed:
fix clog

bcoca committed:
cleanup

bcoca committed:
upd

webknjaz commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -140,94 +153,163 @@
+    def list_installed(self):
+        rc, out, err = module.run_command([self._cli, 'query', "%%%s" % '\t%'.join(['n', 'v', 'R', 't', 'a', 'q', 'o', 'p', 'V'])])
+        if rc != 0 or err:
+            raise Exception("Unable to list packages rc=%s : %s" % (rc, err))
```
Let's use `RuntimeError`
```suggestion
            raise RuntimeError("Unable to list packages rc=%s : %s" % (rc, err))
```

webknjaz commented on lib/ansible/modules/packaging/language/pip_package_info.py:
```diff
@@ -0,0 +1,152 @@
+
+- name: get from specific paths (virtualenvs?)
+  pip_package_info:
+    clients: '/home/me/projec42/python/pip3.5'
```
```suggestion
    clients: '/home/me/project42/python/pip3.5'
```

bcoca committed:
removed pip as that was removed

maxamillion commented:
looks good, needs some tests and then I'll merge.

bcoca committed:
renamed cpan

bcoca committed:
added a single line since 2 lines are needed to be

readabnle instead of just 1 line, it is a huge problem otherwise

bcoca committed:
fix internal ref

webknjaz commented on lib/ansible/module_utils/facts/packages.py:
```diff
@@ -0,0 +1,82 @@
+
+def get_all_pkg_managers():
+
+    return dict([(obj.__name__.lower(), obj) for obj in get_all_subclasses(PkgMgr) if obj not in(CLIMgr, LibMgr)])
```
`in()` looks like a function call, extra list creation not needed (generator expression should be fine) and it's better readable when multiline:
```suggestion
    return dict(
        (obj.__name__.lower(), obj) for obj in get_all_subclasses(PkgMgr)
        if obj not in (CLIMgr, LibMgr)
    )
```

bcoca replied:
fixed in, keeping as list as it is slightly more efficient (always consumed)

webknjaz commented on lib/ansible/module_utils/facts/packages.py:
```diff
@@ -0,0 +1,82 @@
+
+    @abstractmethod
+    def is_available(self):
+        ''' This method is supposed to return True/False if the package manager is currently installed/usable'''
```
Could you please follow PEP 257 for docstrings?

bcoca replied:
done

webknjaz commented on lib/ansible/module_utils/facts/packages.py:
```diff
@@ -0,0 +1,82 @@
+    def get_packages(self):
+        ''' Take all of the above and return a dictionary of lists of dictionaries (package = list of installed versions) '''
+
+        installed_packages = {}
```
how about
```suggestion
        installed_packages = defaultdict(list)
```

bcoca replied:
i don't see why we need to do extra work here

webknjaz commented on lib/ansible/module_utils/facts/packages.py:
```diff
@@ -0,0 +1,82 @@
+        found = False
+        try:
+            self._lib = __import__(self.LIB)
+            found = True
```
just
```suggestion
            return True
```
here, no need for another var

bcoca replied:
i keep the variable as it simplifies adding conditions and having multiple returns makes it harder to update the type and can get lost in the shuffle.

webknjaz commented on lib/ansible/module_utils/facts/packages.py:
```diff
@@ -0,0 +1,82 @@
+    def is_available(self):
+        found = False
+        try:
+            self._lib = __import__(self.LIB)
```
Using this directly is discouraged: https://docs.python.org/3/library/importlib.html#importlib.__import__
> Note Programmatic importing of modules should use import_module() instead of this function.

bcoca replied:
not an option with 2.6

webknjaz commented on lib/ansible/module_utils/facts/packages.py:
```diff
@@ -0,0 +1,82 @@
+        self._lib = None
+        super(LibMgr, self).__init__()
+
+    def is_available(self):
```
Use a more clear name to better reflect what the function is actually doing. It doesn't just check availability but also mutates state.
```suggestion
    def try_import(self):
```

bcoca replied:
generic name that we arrived at during previous reviews , it does can have side effects so i added comment to base class to indicate this

webknjaz commented on lib/ansible/module_utils/facts/packages.py:
```diff
@@ -0,0 +1,82 @@
+        super(CLIMgr, self).__init__()
+
+    def is_available(self):
+        self._cli = get_bin_path(self.CLI, False)
```
Again, this doesn't look like an availability check.

webknjaz commented on lib/ansible/modules/packaging/language/cpan_package_facts.py:
```diff
@@ -0,0 +1,105 @@
+def main():
+
+    # start work
+    global module
```
can we not do this, please? It's not even used in the outer scope...
```suggestion
```

webknjaz commented on lib/ansible/modules/packaging/language/cpan_package_facts.py:
```diff
@@ -0,0 +1,105 @@
+        module.warn('Failed to retrieve packages with %s: %s' % (cpan._cli, to_text(e)))
+
+    if found == 0:
+        module.fail_json(msg='Unable to find a usable cpan')
```
Why can't this just be in the `except` block?

webknjaz commented on lib/ansible/modules/packaging/language/cpan_package_facts.py:
```diff
@@ -0,0 +1,105 @@
+        if cpan.is_available():
+            found += 1
+            packages = cpan.get_packages()
+    except Exception as e:
```
can this be more precise?

bcoca committed:
not intended in this round

bcoca committed:
updated as per fb
